MP-4509 Table Design (#233)
spryker/spryker-docs: docs/marketplace/dev/front-end/table-design/index.md

---
title: Table design
description: This article describes the Table Design in the Components Library.
template: concept-topic-template
---

This article describes the Table Design in the Components Library.

## Overview

A Table Component is an arrangement of data in rows and columns, or possibly in a more complex structure 
(with sorting, filtering, pagination, row selections, infinite scrolling, etc.). 
It is an essential building block of a user interface.

> A basic able Component is `<spy-table [config]="config"></spy-table>` where `config` is:

Reviewer: ```suggestion
A basic table component is `<spy-table [config]="config"></spy-table>` where `config` is:
```

> - `dataSource` - an array of rows to be displayed on the table.
> - `columns` - an array of columns.

Reviewer: ```suggestion
- `dataSource` - Datasource configuration from which the data is taken.
- `columns` - an array of columns configuration.
```

```ts
const config: TableConfig = {
  dataSource: {
    // transforms input data via Data Transformer service
    type: DatasourceType,
    transform?: DataTransformerConfig,
  },
  columns: [
    { id: 'col1', title: 'Column #1' },
    { id: 'col2', title: 'Column #2' },
    { id: 'col3', title: 'Column #3' },
  ],
};
```

## Architecture

![Table Architecture](https://spryker.s3.eu-central-1.amazonaws.com/docs/Marketplace/dev+guides/Front-end/table-architecture.svg)

> ### Components

Reviewer: This section is not clear. What are you trying to explain here?

Author: This is the main section of table architecture with a diagram and subsections like features, components, etc.

    total = {
        enabled: true,
    };
    columnConfigurator = {
        enabled: true,
    };
}

@NgModule({
    imports: [
        TableModule.forRoot(),
        TableFiltersFeatureModule.withFilterComponents({
            select: TableFilterSelectComponent,
            'date-range': TableFilterDateRangeComponent,
            'tree-select': TableFilterTreeSelectComponent,
        } as any),

        // Table Filter Modules
        TableFilterSelectModule,
        TableFilterDateRangeModule,
        TableFilterTreeSelectModule,
    ],
    providers: [
        {
            provide: TableDefaultConfig,
            useClass: TableDefaultConfigData,
        },
    ],
})
export class DefaultTableConfigModule {}
```

### Configuration

The Table Component is configured via so called [Table Configuration](/docs/marketplace/dev/front-end/table-design/table-configuration.html) 
that sets up how table will behave and look like.

### Datasources

To render data Table must receive it via so called [Datasources](/docs/marketplace/dev/front-end/ui-components-library/datasources.html) that is registered by the user and then configured via Table Configuration.

### Features

> Every other piece of functionality was extracted into so called `Table Feature`:

Reviewer: Please also add a link to where all table features are described.

- Table Feature is simply an Angular Component that encapsulates specific extension of the Core Table.
- Core Table contains specific placeholders in it’s view that Table Feature may target to render it’s own piece of UI.
- Most of the common table functionality already exists as a Table Feature and may be used in project.

> To use Features, Table Module must implement a specific interface (TableConfig) and then be registered to the Root Module via TableModule.withFeatures():

Reviewer: ```suggestion
To use a Feature component - register an Angular Module that implements `ModuleWithFeature` interface in the Root Module via TableModule.withFeatures() under the key that will become it's configuration key:
```

```ts

> class TableDefaultConfigData implements Partial<TableConfig> {
>     total = {
>         enabled: true,
>     };
>     columnConfigurator = {
>         enabled: true,
>     };
> }
> 
> @NgModule({
>     imports: [
>         TableModule.forRoot(),
>         TableModule.withFeatures({
>             filters: () => import('@spryker/table.feature.filters').then((m) => m.TableFiltersFeatureModule),
>             pagination: () => import('@spryker/table.feature.pagination').then((m) => m.TablePaginationFeatureModule),
>             rowActions: () => import('@spryker/table.feature.row-actions').then((m) => m.TableRowActionsFeatureModule),
>             search: () => import('@spryker/table.feature.search').then((m) => m.TableSearchFeatureModule),
>             syncStateUrl: () => import('@spryker/table.feature.sync-state').then((m) => m.TableSyncStateFeatureModule),
>             total: () => import('@spryker/table.feature.total').then((m) => m.TableTotalFeatureModule),
>             itemSelection: () =>
>                 import('@spryker/table.feature.selectable').then((m) => m.TableSelectableFeatureModule),
>             batchActions: () =>
>                 import('@spryker/table.feature.batch-actions').then((m) => m.TableBatchActionsFeatureModule),
>             columnConfigurator: () =>
>                 import('@spryker/table.feature.settings').then((m) => m.TableSettingsFeatureModule),
>             title: () => import('@spryker/table.feature.title').then((m) => m.TableTitleFeatureModule),
>             editable: () => import('@spryker/table.feature.editable').then((m) => m.TableEditableFeatureModule),
>         }),
>     ],
>     providers: [
>         {
>             provide: TableDefaultConfig,
>             useClass: TableDefaultConfigData,
>         },
>     ],
> })
> export class DefaultTableConfigModule {}

Reviewer: We do not need all features here just one for example

```suggestion
@NgModule({
    imports: [
        TableModule.withFeatures({
            pagination: () => import('@spryker/table.feature.pagination').then((m) => m.TablePaginationFeatureModule),
        }),
    ],
})
export class AppModule {}
```

```

### Columns

> Columns in a Table are defined by the `Column Type` and are rendered withing the columns (text, image, link, etc.).

Reviewer: Please add a link to all table columns.

New Column Type may be created and registered to the Table.

> To use Column components, Table Module must implement a specific interface (TableConfig) and then be registered to the Root Module via TableModule.withColumnComponents():

Reviewer: ```suggestion
To use Column component it must implement `TableColumn` interface with defined config and then be registered to the Root Module via TableModule.withColumnComponents():
```

```ts

> class TableDefaultConfigData implements Partial<TableConfig> {

Reviewer: Please only keep example of how one column is registered and remove everything else.

    total = {
        enabled: true,
    };
    columnConfigurator = {
        enabled: true,
    };
}

@NgModule({
    imports: [
        TableModule.forRoot(),
        TableModule.withColumnComponents({
            text: TableColumnTextComponent,
            image: TableColumnImageComponent,
            date: TableColumnDateComponent,
            chip: TableColumnChipComponent,
            input: TableColumnInputComponent,
            select: TableColumnSelectComponent,
            dynamic: TableColumnDynamicComponent,
            autocomplete: TableColumnAutocompleteComponent,
        } as any),

        // Table Column Type Modules
        TableColumnChipModule,
        TableColumnTextModule,
        TableColumnImageModule,
        TableColumnDateModule,
        TableColumnInputModule,
        TableColumnSelectModule,
        TableColumnDynamicModule,
        TableColumnAutocompleteModule,
    ],
    providers: [
        {
            provide: TableDefaultConfig,
            useClass: TableDefaultConfigData,
        },
    ],
})
export class DefaultTableConfigModule {}
```

### Actions 

> There is a way to trigger some [Actions](/docs/marketplace/dev/front-end/ui-components-library/actions.html) by the user interaction with the Table.

Reviewer: Please add info that there are Table Features that can trigger actions like Row Actions Feature and Batch Actions Feature.

  beforeColsHeader = 'before-cols-header',
  beforeCols = 'before-cols',
  cell = 'cell',
  afterCols = 'after-cols',
  afterColsHeader = 'after-cols-header',
  afterRows = 'after-rows',
  afterTable = 'after-table',
  bottom = 'bottom',
  hidden = 'hidden',
}

export interface TableRowActionRegistry {
  // Key is action string - value is action options type
}

export type TableRowAction = keyof TableRowActionRegistry;

export interface TableRowActionHandler {
  handleAction(actionEvent: TableActionTriggeredEvent): void;
}

export interface TableRowActionsDeclaration {
  [type: string]: TableRowActionHandler;
}

export interface TableRowClickEvent {
  row: TableDataRow;
  event: Event;
}
```

## Extensions

There are a few common Table extensions that are available in UI library:

- [Column Type](/docs/marketplace/dev/front-end/table-design/table-column-type-extension/) - describes how a specific type of the column is rendered within a table column.
- [Feature](/docs/marketplace/dev/front-end/table-design/table-feature-extension/) - encapsulating a piece of UI that is targeted to a specific location within a Table Component or that may provide additional functionality.

> - [Filter](/docs/marketplace/dev/front-end/table-design/table-filter-extension/) - provides filtering functionality to the Core Table Component.

Reviewer: Please remove this section as everything is described above.
